File: architecture-skills/design-patterns/examples/bridge/bridge.go
// Bridge pattern in Go.
//
// The shape structs hold a Renderer interface value, so shapes and
// renderers form two independent axes. No inheritance is needed: the
// "abstraction hierarchy" is just the types sharing the Shape interface.

package main

import "fmt"

type Renderer interface {
	Circle(radius float64) string
	Rectangle(width, height float64) string
}

type VectorRenderer struct{}

func (VectorRenderer) Circle(radius float64) string {
	return fmt.Sprintf("<circle r=%.1f/>", radius)
}
func (VectorRenderer) Rectangle(width, height float64) string {
	return fmt.Sprintf("<rect w=%.1f h=%.1f/>", width, height)
}

type RasterRenderer struct{}

func (RasterRenderer) Circle(radius float64) string {
	return fmt.Sprintf("pixels for a circle of radius %.1f", radius)
}
func (RasterRenderer) Rectangle(width, height float64) string {
	return fmt.Sprintf("pixels for a %.1fx%.1f rectangle", width, height)
}

type Shape interface {
	Draw() string
}

type Circle struct {
	Renderer Renderer
	Radius   float64
}

func (c Circle) Draw() string { return c.Renderer.Circle(c.Radius) }

type Rectangle struct {
	Renderer      Renderer
	Width, Height float64
}

func (r Rectangle) Draw() string { return r.Renderer.Rectangle(r.Width, r.Height) }

func main() {
	for _, renderer := range []Renderer{VectorRenderer{}, RasterRenderer{}} {
		shapes := []Shape{
			Circle{Renderer: renderer, Radius: 5},
			Rectangle{Renderer: renderer, Width: 3, Height: 4},
		}
		for _, shape := range shapes {
			fmt.Println(shape.Draw())
		}
	}
}

File: architecture-skills/design-patterns/examples/bridge/bridge.swift
// Bridge pattern in Swift.
//
// Shapes hold a renderer as a protocol-typed property, keeping the two
// hierarchies independent. Value-type shapes with a protocol "handle"
// to the implementation are the natural Swift form of this pattern.

import Foundation

protocol Renderer {
    func circle(radius: Double) -> String
    func rectangle(width: Double, height: Double) -> String
}

struct VectorRenderer: Renderer {
    func circle(radius: Double) -> String {
        "<circle r=\(String(format: "%.1f", radius))/>"
    }
    func rectangle(width: Double, height: Double) -> String {
        "<rect w=\(String(format: "%.1f", width)) h=\(String(format: "%.1f", height))/>"
    }
}

struct RasterRenderer: Renderer {
    func circle(radius: Double) -> String {
        "pixels for a circle of radius \(String(format: "%.1f", radius))"
    }
    func rectangle(width: Double, height: Double) -> String {
        "pixels for a \(String(format: "%.1f", width))x\(String(format: "%.1f", height)) rectangle"
    }
}

protocol Shape {
    func draw() -> String
}

struct Circle: Shape {
    let renderer: Renderer
    let radius: Double

    func draw() -> String { renderer.circle(radius: radius) }
}

struct Rectangle: Shape {
    let renderer: Renderer
    let width: Double
    let height: Double

    func draw() -> String { renderer.rectangle(width: width, height: height) }
}

let renderers: [Renderer] = [VectorRenderer(), RasterRenderer()]
for renderer in renderers {
    let shapes: [Shape] = [
        Circle(renderer: renderer, radius: 5.0),
        Rectangle(renderer: renderer, width: 3.0, height: 4.0),
    ]
    for shape in shapes {
        print(shape.draw())
    }
}
